README.md
# Starscape

Starscape is a [Blender](http://www.blender.org) add-on for adding stars to the background of a scene.

![Moon-Earth](https://raw.githubusercontent.com/wiki/mrossini-ethz/starscape/images/moon-earth.png)

## Features
The add-on provides the following features:
- Procedural generation of stars
- Realistic brightness and colour distribution
- Realistic number of stars in the sky (but adjustable)
- Stars are actual objects in the scene (no textures!)
- Star appearance is independent from render resolution and camera focal length (point sources)
- Appearance can be changed by hand after generation
- Compatible with all kinds of world backgrounds

## Blender Artist thread
[Link](https://blenderartists.org/t/starscape-add-on/1336930 "Link") to Blender Artist page

## Installation
1. Download the latest [release](https://github.com/mrossini-ethz/starscape/releases) and save it in a directory of your convenience.
2. Open Blender.
3. In th menu go to Edit -> Preferences -> Addons.
4. At the top of the window, chose *Install*.
5. Select the file downloaded zip file and press *Install Addon*.
6. Search for *Scene: Starscape* in the Addon list.
7. Activate the checkbox for the plugin.
8. If you want to keep the addon activated when blender restarts, open the menu (bottom left menu button) and choose *Save Preferences*.

## Usage

### Addon Panel
The controls for the addon can be found in the Properties panel under World > Starscape.

### Execution
Press 'Generate Starscape' in the panel. That's it! An object called 'Starscape' is created. It is anchored to the camera and thereby appears to remain fixed in the render even when moving the camera.

At the moment there are no options. This will change in the future.

### Options

#### Random Seed
The stars are randomly generated. In order to get reproducible results, the random number generator is initialised with a seed value. If the random result is not pleasing in some way (e.g. because there are a bunch of stars too close together by chance), try changing this value.

#### Star Density
The star density can be different depending where you are within a given galaxy. The density value of 1 is used here to represent the conditions found here near the earth. If you set the density to 2 you will see twice as many stars. Note that increasing this number is not realistic in our part of space. Also note that complexity of the scene increases with this number.

#### Star Intensity
You can change the intensity of the stars using this value. At the moment, the value 1 is arbitrary. Note that the interpretation of the intensity value may change in the future to more realistically render stars with a given film exposure.

#### Hemisphere
For scenes on a planet or moon the stars below the horizon are not necessary. This option reduces the amount of computing resources needed.

#### Lock To Camera
Attaches the starscape object to the camera such that the stars remain fixed with respect to the camera. This should always be activated because stars are usually very far away from an observer and their movement is negligible with respect to the motion of the observer. If you have a star that is very close (such as the sun), you should model it separately.

#### Clear World Background
Clears the world background material and sets the color to pure black.

## Planned features
The following features are planned for the future:
- Increase intensity range to fainter stars
- Add a milky way, i.e. a band of higher star density
- Twinkling stars in animations
- Change the star shape

## Changelog

### Version 0.3
Added options:
- Clear world background

### Version 0.2
Added basic options:
- Random seed
- Star density
- Star intensity
- Hemisphere
- Camera lock

### Version 0.1
First official release.

## License
Camera Calibration with Perspective Views of Rectangles

Copyright (C) 2021  Marco Rossini

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
version 2 as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

This Blender plugin is based on the research paper "Recovery of Intrinsic
and Extrinsic Camera Parameters Using Perspective Views of Rectangles" by
T. N. Tan, G. D. Sullivan and K. D. Baker, Department of Computer Science,
The University of Reading, Berkshire RG6 6AY, UK,
from the Proceedings of the British Machine Vision Conference, published by
the BMVA Press.


ChangeLog
ChangeLog
---------

Version 0.2: Added options
- Clear world background

Version 0.2: Added basic options
- Random seed
- Star density
- Star intensity
- Hemisphere
- Camera lock

Version 0.1: First official release.


operator.py
# Blender Plugin: Starscape
# Copyright (C) 2021  Marco Rossini
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# version 2 as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

# This Blender plugin is based on the research paper "Recovery of Intrinsic
# and Extrinsic Camera Parameters Using Perspective Views of Rectangles" by
# T. N. Tan, G. D. Sullivan and K. D. Baker, Department of Computer Science,
# The University of Reading, Berkshire RG6 6AY, UK, Email: T.Tan@reading.ac.uk,
# from the Proceedings of the British Machine Vision Conference, published by
# the BMVA Press.

import bpy
import mathutils

from . import starscape

class StarscapeOperator(bpy.types.Operator):
    bl_idname = "world.starscape"
    bl_label = "Generate Starscape"
    bl_description = "Generates a background with stars or updates an existing starscape."

    def execute(self, context):
        # Get the properties
        props = bpy.context.scene.starscape_properties

        starscape.generate_starscape(props)

        return {'FINISHED'}


starscape.py
import bpy
import sys
import math
import random
import mathutils

### Helper function ########################################################################

def create_or_reuse_mesh_object(name):
    # Load/create the mesh
    try:
        mesh = bpy.data.meshes[name + "_mesh"]
        mesh.clear_geometry()
    except KeyError:
        mesh = bpy.data.meshes.new(name + "_mesh")
    # Load/create the object
    try:
        obj = bpy.data.objects[name]
    except KeyError:
        obj = bpy.data.objects.new(name, mesh)
        bpy.context.collection.objects.link(obj)
    return obj

def random_spherical_coordinates():
    phi = random.random() * 2 * math.pi
    theta = math.asin(2 * random.random() - 1)
    return phi, theta

def spherical_to_cartesian_coordinates(radius, phi, theta):
    x = radius * math.cos(phi) * math.cos(theta)
    y = radius * math.sin(phi) * math.cos(theta)
    z = radius * math.sin(theta)
    return x, y, z

def hide_node_outputs(node):
    for socket in node.outputs:
        if not socket.is_linked:
            socket.hide = True

def make_node(nodes, node_type, x, y):
    node = nodes.new(node_type)
    node.location = (x, y)
    return node

def make_group_node(nodes, group, x, y):
    node = nodes.new("ShaderNodeGroup")
    node.node_tree = group
    node.location = (x, y)
    return node

def make_math_node(nodes, function, x, y, default_1 = 0.5, default_2 = 0.5):
    node = nodes.new("ShaderNodeMath")
    node.operation = function
    node.location = (x, y)
    node.inputs[0].default_value = default_1
    node.inputs[1].default_value = default_2
    return node

def connect_nodes(tree, *args):
    if len(args) < 4 or (len(args) - 4) % 3 != 0:
        raise Exception("Invalid number of arguments!")

    # Make n links
    n = (len(args) - 1) // 3
    for i in range(n):
        node_a = args[3 * i]
        node_b = args[3 * (i + 1)]
        output = args[3 * i + 1]
        input = args[3 * i + 2]
        tree.links.new(node_b.inputs[input], node_a.outputs[output])

def make_group_inputs(group, x, y, *sockets):
    if len(sockets) % 2 != 0:
        raise Exception("Invalid number of arguments!")
    node = group.nodes.new("NodeGroupInput")
    node.location = (x, y)
    for i in range(len(sockets) // 2):
        group.inputs.new("NodeSocket" + sockets[2 * i], sockets[2 * i + 1])
    return node

def make_group_outputs(group, x, y, *sockets):
    if len(sockets) % 2 != 0:
        raise Exception("Invalid number of arguments!")
    node = group.nodes.new("NodeGroupOutput")
    node.location = (x, y)
    for i in range(len(sockets) // 2):
        group.outputs.new("NodeSocket" + sockets[2 * i], sockets[2 * i + 1])
    return node

### Main function ##########################################################################

def generate_starscape(props):
    # Get the scene camera
    camera = bpy.context.scene.camera
    if not camera or camera.data.type != "PERSP":
        return False

    # Initialize random number generator
    random.seed(props.random_seed)

    # Create stars positions (mesh data consisting only of vertices, radius = 1)
    vertices = []
    N = 1000 * props.star_density
    if props.hemisphere:
        N = N / 2
    for i in range(round(N)):
        # Generate a random location at random across the sky
        phi, theta = random_spherical_coordinates()
        x, y, z = spherical_to_cartesian_coordinates(1, phi, theta)
        if props.hemisphere:
            z = abs(z)
        vertices.append((x, y, z))

    # Load/create the star location mesh
    stars_obj = create_or_reuse_mesh_object("Starscape")
    # Set the mesh to the star data
    stars_obj.data.from_pydata(vertices, [], [])

    # Set vertex normals
    for v in stars_obj.data.vertices:
        v.normal = v.co

    # Add a star template
    # Usa a triangle, because it uses the least amount of vertices and faces
    vertices = []
    s = 0.0002
    q = s * math.sqrt(3)
    vertices.append((+0, 0, +2 * s))
    vertices.append((-q, 0, -1 * s))
    vertices.append((+q, 0, -1 * s))
    template_obj = create_or_reuse_mesh_object("Star_Template")
    template_obj.data.from_pydata(vertices, [(0, 1), (0, 2), (1, 2)], [(0, 1, 2)])

    ### Material ###########################################################################

    # Create the new material "Star Shader"
    try:
        shader = bpy.data.materials["Star Shader"]
    except KeyError:
        shader = bpy.data.materials.new("Star Shader")

    # Use nodes
    shader.use_nodes = True
    node_tree = shader.node_tree
    nodes = node_tree.nodes
    nodes.clear()

    # Add Material Output node
    material_output = make_node(nodes, "ShaderNodeOutputMaterial", 0, 0)

    # Add an Emission Shader node
    emission = make_node(nodes, 'ShaderNodeEmission', -200, 0)

    # Add a math node for multiplication with light path
    math_lightpath = make_math_node(nodes, "MULTIPLY", -400, 0)
    light_path = make_node(nodes, "ShaderNodeLightPath", -600, 100)
    connect_nodes(node_tree, light_path, "Is Camera Ray", 0, math_lightpath)

    # Add a math node to control the intensity
    math_intensity = make_math_node(nodes, "MULTIPLY", -600, 0, default_2 = 15 * props.star_intensity)

    # Add a node group for random intensity
    group = bpy.data.node_groups.new("Random Intensity", "ShaderNodeTree")
    group_inputs = make_group_inputs(group, 0, 0, "Float", "Random")
    group_outputs = make_group_outputs(group, 1200, 0, "Float", "Intensity")
    # Add a math node to prepare random input
    math_inp_fact = make_math_node(group.nodes, "MULTIPLY", 200, 0, default_2 = 9100)
    # Add a math node to prepare input
    math_pre_div = make_math_node(group.nodes, "DIVIDE", 400, 0, default_2 = 3.56)
    # Add a math node to convert visual magnitude
    math_mag_log = make_math_node(group.nodes, "LOGARITHM", 600, 0, default_2 = math.e)
    # Add a math node to convert visual magnitude
    math_mag_div = make_math_node(group.nodes, "DIVIDE", 800, 0, default_2 = -1.21)
    # Add a math node to convert visual magnitude
    math_mag_power = make_math_node(group.nodes, "POWER", 1000, 0, default_1 = 2.512)
    # Connect nodes
    connect_nodes(group, group_inputs, "Random",
        0, math_inp_fact, 0,
        0, math_pre_div, 0,
        0, math_mag_log, 0,
        0, math_mag_div, 0,
        1, math_mag_power, 0,
        "Intensity", group_outputs)
    # Add group
    random_magnitude = make_group_node(nodes, group, -800, 0)

    # Make node group to split single random value into two
    group = bpy.data.node_groups.new("Random Splitter", "ShaderNodeTree")
    group_inputs = make_group_inputs(group, 0, 0, "Float", "Random")
    group_outputs = make_group_outputs(group, 600, 0, "Float", "Random 1", "Float", "Random 2")
    # Multiplication by large factor
    math_rsplit_mult = make_math_node(group.nodes, "MULTIPLY", 200, -100, default_2 = 1000)
    # Use only decimals
    math_rsplit_mod = make_math_node(group.nodes, "MODULO", 400, -100, default_2 = 1)
    # Connect nodes
    connect_nodes(group, group_inputs, "Random", "Random 1", group_outputs)
    connect_nodes(group, group_inputs, "Random",
        0, math_rsplit_mult, 0,
        0, math_rsplit_mod, 0,
        "Random 2", group_outputs)
    # Add group
    random_splitter = make_group_node(nodes, group, -1000, 0)

    # Add geometry input node
    geometry_input = make_node(nodes, "ShaderNodeObjectInfo", -1200, 0)

    # Connect main node chain
    connect_nodes(node_tree, geometry_input, "Random",
        "Random", random_splitter, "Random 1",
        "Random", random_magnitude, "Intensity",
        0, math_intensity, 0,
        1, math_lightpath, 0,
        "Strength", emission, "Emission",
        "Surface", material_output)
    hide_node_outputs(light_path)
    hide_node_outputs(geometry_input)

    # Make node group for star color
    group = bpy.data.node_groups.new("Random Star Color", "ShaderNodeTree")
    group_inputs = make_group_inputs(group, 0, 0, "Float", "Random")
    group_outputs = make_group_outputs(group, 800, 0, "Color", "Color")
    # Color temperature width
    math_kelvin_width = make_math_node(group.nodes, "MULTIPLY", 200, 0, default_2 = 17000)
    # Color temperature offset
    math_kelvin_offset = make_math_node(group.nodes, "ADD", 400, 0, default_2 = 3000)
    # Blackbody color
    blackbody = make_node(group.nodes, "ShaderNodeBlackbody", 600, 0)
    # Connect nodes
    connect_nodes(group, group_inputs, "Random",
        0, math_kelvin_width, 0,
        0, math_kelvin_offset, 0,
        "Temperature", blackbody, "Color",
        "Color", group_outputs)
    # Add group
    random_color = make_group_node(nodes, group, -600, -200)

    # Connect color chain
    connect_nodes(node_tree, random_splitter, "Random 2",
        "Random", random_color, "Color",
        "Color", emission)

    # Set material
    template_obj.active_material = bpy.data.materials["Star Shader"]

    ### Constraints and relationships ######################################################

    # Add location constraint
    # This keeps the stars fixed relatve to the camera
    stars_obj.constraints.clear()
    if props.camera_lock:
        constraint = stars_obj.constraints.new(type="COPY_LOCATION")
        constraint.target = obj_camera = camera

    # Add driver to object scale for the stars
    # This makes the stars be as far away while still visible
    fcurves = stars_obj.driver_add("scale")
    for fcurve in fcurves:
        # Add a variable for the camera focal length
        var = fcurve.driver.variables.new()
        var.name = "s"
        var.type = "SINGLE_PROP"
        target = var.targets[0]
        target.id_type = "CAMERA"
        target.id = camera.data.id_data
        target.data_path = "clip_end"

        # Set the driver expression
        fcurve.driver.expression = "0.9 * s"

    # Add driver to object scale for the template
    # This keeps the star size independent from focal length and render resolution
    fcurves = template_obj.driver_add("scale")
    for fcurve in fcurves:
        # Add a variable for the camera focal length
        var = fcurve.driver.variables.new()
        var.name = "f"
        var.type = "SINGLE_PROP"
        target = var.targets[0]
        target.id_type = "CAMERA"
        target.id = camera.data.id_data
        target.data_path = "lens"

        # Add a variable for the render width
        var = fcurve.driver.variables.new()
        var.name = "x"
        var.type = "SINGLE_PROP"
        target = var.targets[0]
        target.id_type = "SCENE"
        target.id = bpy.context.scene.id_data
        target.data_path = "render.resolution_x"

        # Add a variable for the render height
        var = fcurve.driver.variables.new()
        var.name = "y"
        var.type = "SINGLE_PROP"
        target = var.targets[0]
        target.id_type = "SCENE"
        target.id = bpy.context.scene.id_data
        target.data_path = "render.resolution_y"

        # Add a variable for the resolution percentage
        var = fcurve.driver.variables.new()
        var.name = "p"
        var.type = "SINGLE_PROP"
        target = var.targets[0]
        target.id_type = "SCENE"
        target.id = bpy.context.scene.id_data
        target.data_path = "render.resolution_percentage"

        # Set the driver expression
        fcurve.driver.expression = "50 / f * 2202.907 / max(x, y) / p * 100"

    # Set object relationship
    template_obj.parent = stars_obj
    stars_obj.instance_type = "VERTS"
    stars_obj.use_instance_vertices_rotation = True
    #stars_obj.show_instancer_for_viewport = False
    stars_obj.show_instancer_for_render = False

    # Hide the template
    template_obj.hide_viewport = True
    #template_obj.hide_render = True
    # Add the object

    # Clear the world background
    if props.clear_world_bg:
        # Turn nodes off
        bpy.context.world.use_nodes = False
        # Set color to pure black
        bpy.context.world.color = (0.0, 0.0, 0.0)

    return True

if __name__ == "__main__":
    generate_starscape()


properties.py
import bpy

class StarscapeProperties(bpy.types.PropertyGroup):
    random_seed: bpy.props.IntProperty(name="Random Seed", default=0, min=0, description="Initial value for the random number generator that determines the random outcome of the starscape generation.")
    star_density: bpy.props.FloatProperty(name="Star Density", default=1.0, min=0.0, max=10.0, description="Density of stars relative to near-earth conditions.")
    star_intensity: bpy.props.FloatProperty(name="Star Intensity", default=1.0, min=0.0, description="Brightness of the rendered stars.")
    hemisphere: bpy.props.BoolProperty(name="Hemisphere", default=False, description="Generate only stars above the horizon.")
    camera_lock: bpy.props.BoolProperty(name="Lock To Camera", default=True, description="Locks the starscape object to the camera.")
    clear_world_bg: bpy.props.BoolProperty(name="Clear World Background", default=False, description="Clears the world background material.")


__init__.py
# Blender Plugin: Starscape
# Copyright (C) 2021  Marco Rossini
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# version 2 as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

# This Blender plugin is based on the research paper "Recovery of Intrinsic
# and Extrinsic Camera Parameters Using Perspective Views of Rectangles" by
# T. N. Tan, G. D. Sullivan and K. D. Baker, Department of Computer Science,
# The University of Reading, Berkshire RG6 6AY, UK, Email: T.Tan@reading.ac.uk,
# from the Proceedings of the British Machine Vision Conference, published by
# the BMVA Press.

bl_info = {
    "name": "Starscape",
    "author": "Marco Rossini",
    "version": (0, 2, 0),
    "warning": "This is an unreleased development version.",
    "blender": (2, 90, 0),
    "location": "Properties > World Properties > Starscape",
    "description": "Creates a procedural starscape for use as a scene background",
    "wiki_url": "https://github.com/mrossini-ethz/starscape",
    "tracker_url": "https://github.com/mrossini-ethz/starscape/issues",
    "support": "COMMUNITY",
    "category": "Scene"
}

if "bpy" in locals():
    import importlib as imp
    imp.reload(starscape)
    imp.reload(operator)
    imp.reload(properties)
    imp.reload(panel)
else:
    from . import starscape
    from . import operator
    from . import properties
    from . import panel

import bpy

def register():
    bpy.utils.register_class(properties.StarscapeProperties)
    bpy.types.Scene.starscape_properties = bpy.props.PointerProperty(type=properties.StarscapeProperties)
    bpy.utils.register_class(operator.StarscapeOperator)
    bpy.utils.register_class(panel.StarscapePanel)


def unregister():
    bpy.utils.unregister_class(panel.StarscapePanel)
    bpy.utils.unregister_class(operator.StarscapeOperator)
    bpy.utils.unregister_class(properties.StarscapeProperties)

if __name__ == "__main__":
    register()
